src/OpenLoco/Map/MapGenerator.cpp

```diff
@@ -625,7 +626,79 @@ namespace OpenLoco::Map::MapGenerator
     // 0x004BDA49
     static void generateTrees()
     {
-        call(0x004BDA49);
+        const auto& options = S5::getOptions();
+
+        // Place forests
+        for (auto i = 0; i < options.numberOfForests; ++i)
+        {
+            const auto randRadius = ((gPrng().randNext(255) * std::max(options.maxForestRadius - options.minForestRadius, 0)) / 255 + options.minForestRadius) * tile_size;
+            const auto randLoc = Map::TilePos2(gPrng().randNext(map_rows), gPrng().randNext(map_columns));
+            const auto randDensity = (gPrng().randNext(15) * std::max(options.maxForestDensity - options.minForestDensity, 0)) / 15 + options.minForestDensity;
+            placeTreeCluster(randLoc, randRadius, randDensity, std::nullopt);
+
+            if (TileManager::numFreeElements() < 0x36000)
+            {
+                break;
+            }
+        }
+
+        // Place a number of random trees
+        for (auto i = 0; i < options.numberRandomTrees; ++i)
+        {
+            const auto randLoc = Map::Pos2(gPrng().randNext(map_width), gPrng().randNext(map_height));
+            placeRandomTree(randLoc, std::nullopt);
+        }
+
+        // Cull trees that are too high / low
+        uint32_t randMask = gPrng().randNext();
+        uint32_t i = 0;
+        std::vector<TileElement*> toBeRemoved;
+        for (auto& loc : TilePosRangeView({}, { map_rows - 1, map_columns - 1 }))
+        {
+            auto tile = TileManager::get(loc);
+            for (auto& el : tile)
+            {
+                auto* elTree = el.as<TreeElement>();
+                if (elTree == nullptr)
+                {
+                    continue;
+                }
+
+                if (elTree->baseHeight() / kMicroToSmallZStep <= options.minAltitudeForTrees)
+                {
+                    if (elTree->baseHeight() / kMicroToSmallZStep != options.minAltitudeForTrees
+                        || (randMask & (1 << i)))
+                    {
+                        toBeRemoved.push_back(&el);
+                        i++;
+                        i %= 32;
+                        break;
+                    }
+                }
+
+                if (elTree->baseHeight() / kMicroToSmallZStep >= options.maxAltitudeForTrees)
+                {
+                    if (elTree->baseHeight() / kMicroToSmallZStep != options.maxAltitudeForTrees
+                        || (randMask & (1 << i)))
+                    {
+                        toBeRemoved.push_back(&el);
+                        i++;
+                        i %= 32;
+                        break;
+                    }
+                }
+            }
+
+            // Remove in reverse order to prevent pointer invalidation
+            for (auto elIter = std::rbegin(toBeRemoved); elIter != std::rend(toBeRemoved); ++elIter)
```

Comment:
This is a slightly different way of doing things compared to vanilla but in vanilla it required messing with the iteration pointer which i wanted to avoid. This has the negative that it allocates but as the allocation can be reused I think it is as sensible as we can hope for.